File: sdks/python/sip_protocol/optimizations.py
"""
Chain-Specific Optimizations for SIP Protocol Python SDK.

Provides optimized configurations for different blockchains:
- Solana: Compute unit budgeting, priority fees
- Ethereum/EVM: Gas estimation, L2 optimizations
- Cross-chain cost comparison
"""

from dataclasses import dataclass
from enum import Enum
from typing import Optional, List, Tuple, Dict


class ChainFamily(Enum):
    """Chain family for optimization selection."""
    SOLANA = "solana"
    EVM = "evm"
    NEAR = "near"
    BITCOIN = "bitcoin"
    COSMOS = "cosmos"


class OptimizationProfile(Enum):
    """Optimization profile (cross-chain)."""
    ECONOMY = "economy"     # Lowest fees, may be slower
    STANDARD = "standard"   # Balanced cost/speed
    FAST = "fast"          # Higher fees, faster confirmation
    URGENT = "urgent"      # Maximum priority


@dataclass
class ChainCharacteristics:
    """Chain characteristics for optimization decisions."""
    family: ChainFamily
    block_time: float  # seconds
    has_eip1559: bool
    is_l2: bool
    cost_tier: int  # 1=cheapest, 5=most expensive
    native_token: str


@dataclass
class SolanaComputeBudget:
    """Solana compute budget configuration."""
    units: int
    microlamports_per_cu: int
    total_priority_fee_lamports: int


@dataclass
class EvmGasConfig:
    """EVM gas configuration."""
    gas_limit: int
    max_fee_per_gas: int  # wei
    max_priority_fee_per_gas: int  # wei


@dataclass
class OptimizationResult:
    """Unified optimization result."""
    chain: str
    family: ChainFamily
    solana: Optional[SolanaComputeBudget] = None
    evm: Optional[EvmGasConfig] = None
    recommendations: Optional[List[str]] = None

    def __post_init__(self):
        if self.recommendations is None:
            self.recommendations = []


# ─── Constants ────────────────────────────────────────────────────────────────

SOLANA_DEFAULT_CU = 200_000
SOLANA_MAX_CU = 1_400_000
SOLANA_DEFAULT_PRIORITY_FEE = 1_000  # microlamports per CU

EVM_BASE_GAS_PRICE = 30_000_000_000  # 30 gwei in wei
ONE_GWEI = 1_000_000_000


# ─── Chain Database ────────────────────────────────────────────────────────────

CHAIN_CHARACTERISTICS: Dict[str, ChainCharacteristics] = {
    "solana": ChainCharacteristics(
        family=ChainFamily.SOLANA,
        block_time=0.4,
        has_eip1559=False,
        is_l2=False,
        cost_tier=1,
        native_token="SOL",
    ),
    "ethereum": ChainCharacteristics(
        family=ChainFamily.EVM,
        block_time=12.0,
        has_eip1559=True,
        is_l2=False,
        cost_tier=5,
        native_token="ETH",
    ),
    "arbitrum": ChainCharacteristics(
        family=ChainFamily.EVM,
        block_time=0.25,
        has_eip1559=True,
        is_l2=True,
        cost_tier=2,
        native_token="ETH",
    ),
    "optimism": ChainCharacteristics(
        family=ChainFamily.EVM,
        block_time=2.0,
        has_eip1559=True,
        is_l2=True,
        cost_tier=2,
        native_token="ETH",
    ),
    "base": ChainCharacteristics(
        family=ChainFamily.EVM,
        block_time=2.0,
        has_eip1559=True,
        is_l2=True,
        cost_tier=2,
        native_token="ETH",
    ),
    "polygon": ChainCharacteristics(
        family=ChainFamily.EVM,
        block_time=2.0,
        has_eip1559=True,
        is_l2=True,
        cost_tier=2,
        native_token="MATIC",
    ),
    "bsc": ChainCharacteristics(
        family=ChainFamily.EVM,
        block_time=3.0,
        has_eip1559=False,
        is_l2=False,
        cost_tier=1,
        native_token="BNB",
    ),
    "near": ChainCharacteristics(
        family=ChainFamily.NEAR,
        block_time=1.0,
        has_eip1559=False,
        is_l2=False,
        cost_tier=1,
        native_token="NEAR",
    ),
}


def detect_chain_family(chain_id: str) -> ChainFamily:
    """Detect chain family from chain identifier."""
    normalized = chain_id.lower()

    if "solana" in normalized:
        return ChainFamily.SOLANA
    elif "near" in normalized:
        return ChainFamily.NEAR
    elif "bitcoin" in normalized or "btc" in normalized:
        return ChainFamily.BITCOIN
    elif "cosmos" in normalized or "osmosis" in normalized:
        return ChainFamily.COSMOS
    else:
        return ChainFamily.EVM  # Default


def get_chain_characteristics(chain_id: str) -> ChainCharacteristics:
    """Get chain characteristics."""
    lower_id = chain_id.lower()

    # Try direct lookup
    if lower_id in CHAIN_CHARACTERISTICS:
        return CHAIN_CHARACTERISTICS[lower_id]

    # Try base name
    base = lower_id.split("-")[0]
    if base in CHAIN_CHARACTERISTICS:
        return CHAIN_CHARACTERISTICS[base]

    # Default
    return ChainCharacteristics(
        family=detect_chain_family(chain_id),
        block_time=12.0,
        has_eip1559=True,
        is_l2=False,
        cost_tier=3,
        native_token="ETH",
    )


# ─── Solana Optimization ──────────────────────────────────────────────────────


def calculate_solana_budget(
    estimated_cu: int,
    profile: OptimizationProfile = OptimizationProfile.STANDARD,
    current_median_fee: Optional[int] = None,
) -> SolanaComputeBudget:
    """Calculate Solana compute budget configuration."""
    # Add 20% buffer
    units = min(int(estimated_cu * 1.2), SOLANA_MAX_CU)

    # Profile multipliers
    multipliers = {
        OptimizationProfile.ECONOMY: 0.5,
        OptimizationProfile.STANDARD: 1.0,
        OptimizationProfile.FAST: 2.0,
        OptimizationProfile.URGENT: 5.0,
    }
    multiplier = multipliers[profile]

    base_fee = current_median_fee or SOLANA_DEFAULT_PRIORITY_FEE
    microlamports_per_cu = max(int(base_fee * multiplier), 100)

    total_priority_fee_lamports = (units * microlamports_per_cu) // 1_000_000

    return SolanaComputeBudget(
        units=units,
        microlamports_per_cu=microlamports_per_cu,
        total_priority_fee_lamports=total_priority_fee_lamports,
    )


def estimate_solana_privacy_cu(
    transfer_count: int,
    creates_atas: bool = False,
    includes_memo: bool = True,
) -> int:
    """Estimate compute units for Solana privacy transaction."""
    cu = 5_000  # Base overhead
    cu += 300  # Compute budget instructions

    # Per transfer
    per_transfer = 35_000 if creates_atas else 10_000
    cu += per_transfer * transfer_count

    # Memo
    if includes_memo:
        cu += 500

    # Key derivation
    cu += 2_000

    return cu


# ─── EVM Optimization ─────────────────────────────────────────────────────────


def calculate_evm_gas(
    estimated_gas: int,
    profile: OptimizationProfile = OptimizationProfile.STANDARD,
    base_fee: Optional[int] = None,
) -> EvmGasConfig:
    """Calculate EVM gas configuration."""
    base = base_fee or EVM_BASE_GAS_PRICE

    # Profile multipliers
    multipliers = {
        OptimizationProfile.ECONOMY: 0.8,
        OptimizationProfile.STANDARD: 1.0,
        OptimizationProfile.FAST: 1.5,
        OptimizationProfile.URGENT: 2.5,
    }
    multiplier = multipliers[profile]

    base_priority = 2 * ONE_GWEI  # 2 gwei
    max_priority_fee_per_gas = int(base_priority * multiplier)
    max_fee_per_gas = base * 2 + max_priority_fee_per_gas

    # 20% buffer
    gas_limit = int(estimated_gas * 1.2)

    return EvmGasConfig(
        gas_limit=gas_limit,
        max_fee_per_gas=max_fee_per_gas,
        max_priority_fee_per_gas=max_priority_fee_per_gas,
    )


def estimate_evm_privacy_gas(
    transfer_count: int,
    includes_approval: bool = False,
    includes_announcement: bool = True,
) -> int:
    """Estimate gas for EVM privacy transaction."""
    gas = 21_000  # Base tx

    # ERC-20 transfers
    gas += 65_000 * transfer_count

    # Approval
    if includes_approval:
        gas += 46_000

    # Announcement
    if includes_announcement:
        gas += 80_000

    return gas


# ─── Unified Optimization ─────────────────────────────────────────────────────


def select_optimal_config(
    chain_id: str,
    profile: OptimizationProfile = OptimizationProfile.STANDARD,
    complexity: str = "medium",
) -> OptimizationResult:
    """
    Select optimal configuration based on chain.

    Args:
        chain_id: Target chain identifier
        profile: Optimization profile
        complexity: Transaction complexity ("simple", "medium", "complex")

    Returns:
        Unified optimization result
    """
    characteristics = get_chain_characteristics(chain_id)
    recommendations = []

    solana_config = None
    evm_config = None

    if characteristics.family == ChainFamily.SOLANA:
        cu_map = {"simple": 50_000, "medium": 150_000, "complex": 300_000}
        estimated_cu = cu_map.get(complexity, 150_000)
        solana_config = calculate_solana_budget(estimated_cu, profile)

        recommendations.append(
            "Solana: Use versioned transactions for complex operations"
        )
        if characteristics.cost_tier == 1:
            recommendations.append(
                "Solana: Very low cost - prioritize speed over savings"
            )

    elif characteristics.family == ChainFamily.EVM:
        gas_map = {"simple": 50_000, "medium": 150_000, "complex": 500_000}
        estimated_gas = gas_map.get(complexity, 150_000)
        evm_config = calculate_evm_gas(estimated_gas, profile)

        if characteristics.is_l2:
            recommendations.append(
                "L2: Lower fees, optimize calldata for L1 data costs"
            )

        if chain_id.lower() == "bsc":
            recommendations.append("BSC: Very low gas costs - use standard profile")

    else:
        recommendations.append(f"Chain {chain_id} not fully optimized yet")

    if characteristics.cost_tier >= 4:
        recommendations.append(
            f"High cost chain (tier {characteristics.cost_tier}) - consider L2 alternatives"
        )

    return OptimizationResult(
        chain=chain_id,
        family=characteristics.family,
        solana=solana_config,
        evm=evm_config,
        recommendations=recommendations,
    )


# ─── Cost Comparison ──────────────────────────────────────────────────────────


def compare_chain_costs(
    chains: List[str],
) -> List[Tuple[str, int, str]]:
    """
    Compare costs across chains.

    Returns:
        List of (chain, cost_tier, recommendation) sorted by cost
    """
    results = []
    for chain in chains:
        chars = get_chain_characteristics(chain)
        recommendations = {
            1: "Excellent - very low costs",
            2: "Good - affordable for frequent use",
            3: "Moderate - suitable for medium value txs",
            4: "Expensive - use for high value only",
            5: "Very expensive - consider alternatives",
        }
        rec = recommendations.get(chars.cost_tier, "Unknown")
        results.append((chain, chars.cost_tier, rec))

    return sorted(results, key=lambda x: x[1])


def recommend_cheapest_chain(
    chains: List[str],
    max_block_time: Optional[float] = None,
) -> Optional[str]:
    """Recommend cheapest viable chain."""
    viable = [
        chain
        for chain in chains
        if max_block_time is None
        or get_chain_characteristics(chain).block_time <= max_block_time
    ]

    if not viable:
        return None

    return min(
        viable,
        key=lambda c: get_chain_characteristics(c).cost_tier,
    )

File: sdks/rust/src/optimizations.rs
//! Chain-Specific Optimizations for SIP Protocol Rust SDK.
//!
//! Provides optimized configurations for different blockchains:
//! - Solana: Compute unit budgeting, priority fees
//! - Ethereum/EVM: Gas estimation, L2 optimizations
//! - Cross-chain cost comparison

use std::collections::HashMap;

// ─── Types ────────────────────────────────────────────────────────────────────

/// Chain family for optimization selection
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ChainFamily {
    Solana,
    Evm,
    Near,
    Bitcoin,
    Cosmos,
}

/// Optimization profile (cross-chain)
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OptimizationProfile {
    /// Lowest fees, may be slower
    Economy,
    /// Balanced cost/speed
    Standard,
    /// Higher fees, faster confirmation
    Fast,
    /// Maximum priority
    Urgent,
}

impl Default for OptimizationProfile {
    fn default() -> Self {
        Self::Standard
    }
}

/// Chain characteristics for optimization decisions
#[derive(Debug, Clone)]
pub struct ChainCharacteristics {
    /// Chain family
    pub family: ChainFamily,
    /// Average block time in seconds
    pub block_time: f64,
    /// Whether chain has EIP-1559 style gas
    pub has_eip1559: bool,
    /// Whether chain is L2/rollup
    pub is_l2: bool,
    /// Relative cost tier (1=cheapest, 5=most expensive)
    pub cost_tier: u8,
    /// Native token symbol
    pub native_token: String,
}

/// Solana compute budget configuration
#[derive(Debug, Clone)]
pub struct SolanaComputeBudget {
    /// Compute units to request
    pub units: u32,
    /// Priority fee in microlamports per CU
    pub microlamports_per_cu: u64,
    /// Total priority fee in lamports
    pub total_priority_fee_lamports: u64,
}

/// EVM gas configuration
#[derive(Debug, Clone)]
pub struct EvmGasConfig {
    /// Gas limit
    pub gas_limit: u64,
    /// Max fee per gas (wei)
    pub max_fee_per_gas: u128,
    /// Max priority fee per gas (wei)
    pub max_priority_fee_per_gas: u128,
}

/// Unified optimization result
#[derive(Debug, Clone)]
pub struct OptimizationResult {
    /// Chain being optimized for
    pub chain: String,
    /// Chain family
    pub family: ChainFamily,
    /// Solana-specific config (if Solana)
    pub solana: Option<SolanaComputeBudget>,
    /// EVM-specific config (if EVM)
    pub evm: Option<EvmGasConfig>,
    /// Recommendations
    pub recommendations: Vec<String>,
}

// ─── Constants ────────────────────────────────────────────────────────────────

/// Default Solana compute units
pub const SOLANA_DEFAULT_CU: u32 = 200_000;

/// Maximum Solana compute units
pub const SOLANA_MAX_CU: u32 = 1_400_000;

/// Default Solana priority fee (microlamports per CU)
pub const SOLANA_DEFAULT_PRIORITY_FEE: u64 = 1_000;

/// EVM base gas price (30 gwei in wei)
pub const EVM_BASE_GAS_PRICE: u128 = 30_000_000_000;

/// One gwei in wei
pub const ONE_GWEI: u128 = 1_000_000_000;

// ─── Chain Detection ──────────────────────────────────────────────────────────

/// Detect chain family from chain identifier
pub fn detect_chain_family(chain_id: &str) -> ChainFamily {
    let normalized = chain_id.to_lowercase();

    if normalized.contains("solana") {
        ChainFamily::Solana
    } else if normalized.contains("near") {
        ChainFamily::Near
    } else if normalized.contains("bitcoin") || normalized.contains("btc") {
        ChainFamily::Bitcoin
    } else if normalized.contains("cosmos") || normalized.contains("osmosis") {
        ChainFamily::Cosmos
    } else {
        // Default to EVM (most common)
        ChainFamily::Evm
    }
}

/// Get chain characteristics
pub fn get_chain_characteristics(chain_id: &str) -> ChainCharacteristics {
    let family = detect_chain_family(chain_id);
    let normalized = chain_id.to_lowercase();

    match normalized.as_str() {
        "solana" | "solana-mainnet" | "solana-devnet" => ChainCharacteristics {
            family: ChainFamily::Solana,
            block_time: 0.4,
            has_eip1559: false,
            is_l2: false,
            cost_tier: 1,
            native_token: "SOL".to_string(),
        },
        "ethereum" | "mainnet" => ChainCharacteristics {
            family: ChainFamily::Evm,
            block_time: 12.0,
            has_eip1559: true,
            is_l2: false,
            cost_tier: 5,
            native_token: "ETH".to_string(),
        },
        "arbitrum" => ChainCharacteristics {
            family: ChainFamily::Evm,
            block_time: 0.25,
            has_eip1559: true,
            is_l2: true,
            cost_tier: 2,
            native_token: "ETH".to_string(),
        },
        "optimism" | "base" => ChainCharacteristics {
            family: ChainFamily::Evm,
            block_time: 2.0,
            has_eip1559: true,
            is_l2: true,
            cost_tier: 2,
            native_token: "ETH".to_string(),
        },
        "bsc" | "bnb" => ChainCharacteristics {
            family: ChainFamily::Evm,
            block_time: 3.0,
            has_eip1559: false,
            is_l2: false,
            cost_tier: 1,
            native_token: "BNB".to_string(),
        },
        "polygon" => ChainCharacteristics {
            family: ChainFamily::Evm,
            block_time: 2.0,
            has_eip1559: true,
            is_l2: true,
            cost_tier: 2,
            native_token: "MATIC".to_string(),
        },
        "near" | "near-mainnet" => ChainCharacteristics {
            family: ChainFamily::Near,
            block_time: 1.0,
            has_eip1559: false,
            is_l2: false,
            cost_tier: 1,
            native_token: "NEAR".to_string(),
        },
        _ => ChainCharacteristics {
            family,
            block_time: 12.0,
            has_eip1559: true,
            is_l2: false,
            cost_tier: 3,
            native_token: "ETH".to_string(),
        },
    }
}

// ─── Solana Optimization ──────────────────────────────────────────────────────

/// Calculate Solana compute budget
///
/// # Arguments
///
/// * `estimated_cu` - Estimated compute units needed
/// * `profile` - Optimization profile
/// * `current_median_fee` - Current median priority fee (optional)
///
/// # Returns
///
/// Compute budget configuration
pub fn calculate_solana_budget(
    estimated_cu: u32,
    profile: OptimizationProfile,
    current_median_fee: Option<u64>,
) -> SolanaComputeBudget {
    // Add 20% buffer
    let units = std::cmp::min((estimated_cu as f64 * 1.2) as u32, SOLANA_MAX_CU);

    // Profile multipliers
    let multiplier = match profile {
        OptimizationProfile::Economy => 0.5,
        OptimizationProfile::Standard => 1.0,
        OptimizationProfile::Fast => 2.0,
        OptimizationProfile::Urgent => 5.0,
    };

    let base_fee = current_median_fee.unwrap_or(SOLANA_DEFAULT_PRIORITY_FEE);
    let microlamports_per_cu = std::cmp::max((base_fee as f64 * multiplier) as u64, 100);

    let total_priority_fee_lamports = (units as u64 * microlamports_per_cu) / 1_000_000;

    SolanaComputeBudget {
        units,
        microlamports_per_cu,
        total_priority_fee_lamports,
    }
}

/// Estimate compute units for privacy transaction
pub fn estimate_solana_privacy_cu(
    transfer_count: u32,
    creates_atas: bool,
    includes_memo: bool,
) -> u32 {
    let mut cu = 5_000; // Base overhead
    cu += 300; // Compute budget instructions

    // Per transfer
    let per_transfer = if creates_atas { 35_000 } else { 10_000 };
    cu += per_transfer * transfer_count;

    // Memo
    if includes_memo {
        cu += 500;
    }

    // Key derivation
    cu += 2_000;

    cu
}

// ─── EVM Optimization ─────────────────────────────────────────────────────────

/// Calculate EVM gas configuration
///
/// # Arguments
///
/// * `estimated_gas` - Estimated gas needed
/// * `profile` - Optimization profile
/// * `base_fee` - Current base fee (wei)
///
/// # Returns
///
/// Gas configuration
pub fn calculate_evm_gas(
    estimated_gas: u64,
    profile: OptimizationProfile,
    base_fee: Option<u128>,
) -> EvmGasConfig {
    let base = base_fee.unwrap_or(EVM_BASE_GAS_PRICE);

    // Profile multipliers for priority fee
    let priority_multiplier = match profile {
        OptimizationProfile::Economy => 0.8,
        OptimizationProfile::Standard => 1.0,
        OptimizationProfile::Fast => 1.5,
        OptimizationProfile::Urgent => 2.5,
    };

    let base_priority = 2 * ONE_GWEI; // 2 gwei base
    let max_priority_fee_per_gas = (base_priority as f64 * priority_multiplier) as u128;
    let max_fee_per_gas = base * 2 + max_priority_fee_per_gas;

    // 20% buffer on gas limit
    let gas_limit = (estimated_gas as f64 * 1.2) as u64;

    EvmGasConfig {
        gas_limit,
        max_fee_per_gas,
        max_priority_fee_per_gas,
    }
}

/// Estimate gas for EVM privacy transaction
pub fn estimate_evm_privacy_gas(
    transfer_count: u32,
    includes_approval: bool,
    includes_announcement: bool,
) -> u64 {
    let mut gas: u64 = 21_000; // Base tx

    // ERC-20 transfers
    gas += 65_000 * transfer_count as u64;

    // Approval
    if includes_approval {
        gas += 46_000;
    }

    // Announcement
    if includes_announcement {
        gas += 80_000;
    }

    gas
}

// ─── Unified Optimization ─────────────────────────────────────────────────────

/// Select optimal configuration based on chain
///
/// # Arguments
///
/// * `chain_id` - Target chain identifier
/// * `profile` - Optimization profile
/// * `complexity` - Transaction complexity ("simple", "medium", "complex")
///
/// # Returns
///
/// Unified optimization result
pub fn select_optimal_config(
    chain_id: &str,
    profile: OptimizationProfile,
    complexity: &str,
) -> OptimizationResult {
    let characteristics = get_chain_characteristics(chain_id);
    let mut recommendations = Vec::new();

    let (solana, evm) = match characteristics.family {
        ChainFamily::Solana => {
            let estimated_cu = match complexity {
                "simple" => 50_000,
                "complex" => 300_000,
                _ => 150_000,
            };
            let budget = calculate_solana_budget(estimated_cu, profile, None);

            recommendations.push("Solana: Use versioned transactions for complex operations".to_string());
            if characteristics.cost_tier == 1 {
                recommendations.push("Solana: Very low cost - prioritize speed over savings".to_string());
            }

            (Some(budget), None)
        }
        ChainFamily::Evm => {
            let estimated_gas = match complexity {
                "simple" => 50_000,
                "complex" => 500_000,
                _ => 150_000,
            };
            let config = calculate_evm_gas(estimated_gas as u64, profile, None);

            if characteristics.is_l2 {
                recommendations.push("L2: Lower fees, optimize calldata for L1 data costs".to_string());
            }

            if chain_id == "bsc" {
                recommendations.push("BSC: Very low gas costs - use standard profile".to_string());
            }

            (None, Some(config))
        }
        _ => {
            recommendations.push(format!("Chain {} not fully optimized yet", chain_id));
            (None, None)
        }
    };

    if characteristics.cost_tier >= 4 {
        recommendations.push(format!(
            "High cost chain (tier {}) - consider L2 alternatives",
            characteristics.cost_tier
        ));
    }

    OptimizationResult {
        chain: chain_id.to_string(),
        family: characteristics.family,
        solana,
        evm,
        recommendations,
    }
}

// ─── Cost Comparison ──────────────────────────────────────────────────────────

/// Compare costs across chains
pub fn compare_chain_costs(chains: &[&str]) -> Vec<(String, u8, String)> {
    let mut results: Vec<_> = chains
        .iter()
        .map(|chain| {
            let chars = get_chain_characteristics(chain);
            let recommendation = match chars.cost_tier {
                1 => "Excellent - very low costs",
                2 => "Good - affordable for frequent use",
                3 => "Moderate - suitable for medium value txs",
                4 => "Expensive - use for high value only",
                _ => "Very expensive - consider alternatives",
            };
            (chain.to_string(), chars.cost_tier, recommendation.to_string())
        })
        .collect();

    results.sort_by_key(|(_, tier, _)| *tier);
    results
}

/// Recommend cheapest viable chain
pub fn recommend_cheapest_chain(
    chains: &[&str],
    max_block_time: Option<f64>,
) -> Option<String> {
    let viable: Vec<_> = chains
        .iter()
        .filter(|chain| {
            let chars = get_chain_characteristics(chain);
            max_block_time.map_or(true, |max| chars.block_time <= max)
        })
        .collect();

    viable
        .into_iter()
        .min_by_key(|chain| get_chain_characteristics(chain).cost_tier)
        .map(|s| s.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_detect_chain_family() {
        assert_eq!(detect_chain_family("solana"), ChainFamily::Solana);
        assert_eq!(detect_chain_family("ethereum"), ChainFamily::Evm);
        assert_eq!(detect_chain_family("arbitrum"), ChainFamily::Evm);
        assert_eq!(detect_chain_family("near"), ChainFamily::Near);
        assert_eq!(detect_chain_family("bitcoin"), ChainFamily::Bitcoin);
    }

    #[test]
    fn test_solana_budget() {
        let budget = calculate_solana_budget(100_000, OptimizationProfile::Standard, None);
        assert_eq!(budget.units, 120_000); // 20% buffer
        assert!(budget.microlamports_per_cu >= 100);
    }

    #[test]
    fn test_evm_gas() {
        let config = calculate_evm_gas(100_000, OptimizationProfile::Standard, None);
        assert_eq!(config.gas_limit, 120_000); // 20% buffer
        assert!(config.max_fee_per_gas > config.max_priority_fee_per_gas);
    }

    #[test]
    fn test_select_optimal_config() {
        let result = select_optimal_config("solana", OptimizationProfile::Standard, "medium");
        assert_eq!(result.family, ChainFamily::Solana);
        assert!(result.solana.is_some());
        assert!(result.evm.is_none());
    }

    #[test]
    fn test_compare_chain_costs() {
        let costs = compare_chain_costs(&["ethereum", "solana", "arbitrum", "bsc"]);
        // Solana and BSC should be cheapest (tier 1)
        assert!(costs[0].1 <= 2);
    }

    #[test]
    fn test_recommend_cheapest() {
        let chains = ["ethereum", "solana", "arbitrum", "bsc"];
        let cheapest = recommend_cheapest_chain(&chains, None);
        assert!(cheapest.is_some());
        // Should be solana or bsc (both tier 1)
        let result = cheapest.unwrap();
        assert!(result == "solana" || result == "bsc");
    }
}
